kai/llm_interfacing/model_provider.py

```diff
@@ -17,201 +24,136 @@
 from pydantic.v1.utils import deep_update
 
 from kai.cache import Cache, CachePathResolver, SimplePathResolver
-from kai.kai_config import KaiConfigModels
+from kai.kai_config import KaiConfigModels, SupportedModelProviders
 from kai.logging.logging import get_logger
 
 LOG = get_logger(__name__)
 tracer = trace.get_tracer("model_provider")
 
 
 class ModelProvider:
+    @staticmethod
+    def from_config(
+        config: KaiConfigModels,
+        demo_mode: bool = False,
+        cache: Cache | None = None,
+    ) -> "ModelProvider":
+        match config.provider:
+            case SupportedModelProviders.CHAT_OLLAMA:
+                return ModelProviderChatOllama(config, demo_mode, cache)
+            case SupportedModelProviders.CHAT_OPENAI:
+                return ModelProviderChatOpenAI(config, demo_mode, cache)
+            case SupportedModelProviders.CHAT_BEDROCK:
+                return ModelProviderChatBedrock(config, demo_mode, cache)
+            case SupportedModelProviders.FAKE_LIST_CHAT_MODEL:
+                return ModelProviderFakeListChatModel(config, demo_mode, cache)
+            case SupportedModelProviders.CHAT_GOOGLE_GENERATIVE_AI:
+                return ModelProviderChatGoogleGenerativeAI(config, demo_mode, cache)
+            case SupportedModelProviders.AZURE_CHAT_OPENAI:
+                return ModelProviderAzureChatOpenAI(config, demo_mode, cache)
+            case SupportedModelProviders.CHAT_DEEP_SEEK:
+                return ModelProviderChatDeepSeek(config, demo_mode, cache)
+            case _:
+                assert_never(config.provider)
```

Comment:
This should be a typed exception rather than an assertion, unless it actually can't be reached. I'm assuming we can trigger this with bad configuration though right?

Reply:
This is a new thing I recently found out about type checking in Python. If a type checker thinks that something can't happen, it will be assigned the type `Never`. This allows you to do [exhaustiveness checking](https://typing.readthedocs.io/en/latest/guides/unreachable.html) and get a static error if you don't handle every case.

We shouldn't ever have an issue with config because the Pydantic model makes sure it's one of the enum values.